evm_loader/program/src/executor/state.rs

```diff
@@ -285,17 +285,39 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
     }
 
     async fn code_hash(&self, from_address: Address, chain_id: u64) -> Result<[u8; 32]> {
-        use solana_program::keccak::hash;
-
-        for action in &self.actions {
-            if let Action::EvmSetCode { address, code, .. } = action {
-                if &from_address == address {
-                    return Ok(hash(code).to_bytes());
-                }
-            }
+        // https://eips.ethereum.org/EIPS/eip-1052
+        // https://eips.ethereum.org/EIPS/eip-161
+
+        #[maybe_async]
+        async fn data_account_exists<B: AccountStorage>(
+            state: &ExecutorState<'_, B>,
+            address: Address,
+            chain_id: u64,
+        ) -> Result<bool> {
+            Ok(state.nonce(address, chain_id).await? > 0
+                || state.balance(address, chain_id).await? > 0)
         }
 
-        Ok(self.backend.code_hash(from_address, chain_id).await)
+        // FIXME: Can we modify self.code to return Option<Buffer> or Option<&[u8]>?
```

Comment:
I agree, we can change self.code to return Option<Buffer>

Reply:
We still need `Inner::Empty`. What we do not need is to store `Option<Buffer>` anywhere. You only care that it is None for codehash logic. Everywhere else it should be immediately unwrapped to the default.

Reply:
I think we should just use the `Buffer` as it is and check for emptiness with `.is_empty()`.

Making it non-empty would only complicate types everywhere with no real benefit. Just like `Vec<T>` and `String`s can be empty, `Buffer` should too.

Reply:
The need of `Inner::Empty` will be resolved here https://github.com/neonlabsorg/neon-evm/pull/259